subprojects/model-core/src/main/java/org/gradle/internal/reflect/MethodSet.java

```diff
@@ -16,38 +16,64 @@
 
 package org.gradle.internal.reflect;
 
+import com.google.common.collect.Maps;
+
 import java.lang.reflect.Method;
-import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.Collection;
 import java.util.Iterator;
-import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static java.util.Collections.unmodifiableCollection;
 
 public class MethodSet implements Iterable<Method> {
-    private final List<Method> methods = new ArrayList<Method>();
-
-    /**
-     * @return true if the method was added, false if not
-     */
-    public boolean add(Method method) {
-        for (Method current : methods) {
-            if (current.getName().equals(method.getName()) && current.getReturnType().equals(method.getReturnType()) && Arrays.equals(current.getParameterTypes(), method.getParameterTypes())) {
-                return false;
-            }
-        }
-        methods.add(method);
-        return true;
+    private final Map<MethodKey, Method> methods = Maps.newLinkedHashMap();
+
+    public void add(Method method) {
+        methods.putIfAbsent(new MethodKey(method), method);
```

Comment:
* the key assures that only the same methods are checked, as before
* putIfAbsent vs put -> previously the first duplicate was kept (e.g. from superclass)